### ISSUE TITLE
Connection pool leases disconnected connections when used with kill-safe-connection 

### ISSUE THREAD

**Issue by sctb** [racket/racket#957]
I submitted the email below to the user@ list, but after testing
further I discovered that the behavior that I desired could be 
obtained by avoiding the use of `kill-safe-connection`.

This might be caused by the use of `last-connected?` by the
`kill-safe-connection%` class in connect-util.rkt, which could 
misrepresent the current state of the proxied connection.

-- 

Hi folks,

We’re using the Racket database library with PostgreSQL and
are looking for a way to manage connections in the presence of
network or database interrupts. Here’s how we’re connecting:

``` 
 (virtual-connection
  (connection-pool
   (lambda ()
     (kill-safe-connection
      (postgresql-connect #:server server
                          #:database database
                          #:user user)))))
```

When restarting the database with active connections, queries
fail with a variety of network exceptions, but low-level connection
objects in the pool remain in a disconnected (raise a “query-exec:
not connected” exception) state even when the database has
recovered. 

What’s the best approach to take to make the connection pool 
more resilient? 

Thanks,

   Scott

**Comment by sctb** [comment]
Indeed, in my local testing, forwarding the `connected?` method
to the connection in the `kill-safe-connection%` proxy resolves 
the issue with subsequent queries throwing the "not connected"
exception.

**Comment by rmculpepper** [comment]
I've pushed a fix to the racket/db repo (racket/db@66044af5a58a41252f4698a2029769b2a903f6fd). The main problem was that `last-connected?` was not getting updated on error; the fix for that was to change the `begin0` to a `dynamic-wind`. The `connected?` method must not block, so I made it forward only if the connection is not in the middle of another method; otherwise, it uses `last-connected?`. Let me know if you still see problems.
